File: src/main/java/org/fraczak/k/avro/Libs.java
package org.fraczak.k.avro;

import org.apache.avro.Schema;
import org.apache.avro.Schema.Field;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

public class Libs {

    public static final ObjectMapper objectMapper = new ObjectMapper();

    public static JsonNode csmToAvro(String typeName, JsonNode csmType) {
        ObjectNode avroSchema = Libs.objectMapper.createObjectNode();
        
        avroSchema.put("name",typeName);
        avroSchema.put("type","record");

        ArrayNode fields = objectMapper.createArrayNode();
        ObjectNode theField = objectMapper.createObjectNode();
        fields.add(theField);
        avroSchema.set("fields",fields);

        if (csmType.has("union")) {
            theField.put("name","union");
            ArrayNode unionType = objectMapper.createArrayNode();

            JsonNode union = csmType.get("union");
            union.fieldNames().forEachRemaining(tagName -> {
                ObjectNode unionTypeRecord = objectMapper.createObjectNode();
                unionTypeRecord.put("name", typeName + "_" + toHex(tagName));
                unionTypeRecord.put("type", "record");
                ArrayNode unionRecordFields = objectMapper.createArrayNode();
                ObjectNode unionRecordField = objectMapper.createObjectNode();
                unionRecordField.put("name","H_" + toHex(tagName));
                unionRecordField.put("type", union.get(tagName).asText());
                unionRecordFields.add(unionRecordField);
                unionTypeRecord.set("fields",unionRecordFields);
                unionType.add(unionTypeRecord);
            });

            theField.set("type",unionType);

        } else if (csmType.has("product")) {
            theField.put("name","product");

            ObjectNode productType = Libs.objectMapper.createObjectNode();
            productType.put("name", typeName + "_record");
            productType.put("type", "record");

            ArrayNode productTypeFields = Libs.objectMapper.createArrayNode();

            JsonNode product = csmType.get("product");
            product.fieldNames().forEachRemaining(tagName -> {
                ObjectNode fieldSpec = objectMapper.createObjectNode();
                fieldSpec.put("name", "H_" + toHex(tagName));
                fieldSpec.put("type", product.get(tagName).asText());
                productTypeFields.add(fieldSpec);
            });
            
            productType.set("fields",productTypeFields);
            theField.set("type",productType);

        } else if (csmType.has("vector")) {
            theField.put("name","vector");

            ObjectNode vectorType = Libs.objectMapper.createObjectNode();
            vectorType.put("name", typeName + "_vector");
            vectorType.put("type", "record");

            ArrayNode vectorTypeFields = Libs.objectMapper.createArrayNode();
            ObjectNode vectorTypeField = objectMapper.createObjectNode();
            vectorTypeField.put("name", "member");
            vectorTypeField.put("type", csmType.get("vector").asText());
            vectorTypeFields.add(vectorTypeField);
            vectorType.set("fields", vectorTypeFields);

            theField.set("type",vectorType);
        } else {
            throw new IllegalArgumentException("Invalid CSM code: " + csmType);
        }

        // System.out.println(avroSchema.toPrettyString());

        return avroSchema;
    }

    
    public static JsonNode avroToCsm(Schema type) {
        Field theField = type.getFields().get(0);

        ObjectNode csmType = objectMapper.createObjectNode();

        switch (theField.name()) {
            case "union": 
                var union = objectMapper.createObjectNode();
                var unionArray = theField.schema().getTypes();
                unionArray.forEach(variant -> {
                    Field field = variant.getFields().get(0);
                    String variantName = fromHex(field.name().substring(2));
                    union.put(variantName, field.schema().getName());
                });
                csmType.set("union", union);
                break;
            case "product":
                var product = objectMapper.createObjectNode();
                var fields = theField.schema().getFields();
                fields.forEach(field -> {
                    String tagName = fromHex(field.name().substring(2));
                    product.put(tagName, field.schema().getName());
                });
                csmType.set("product", product);
                break;
            case "vector":
                String memberType = theField.schema().getField("member").schema().getName();
                csmType.put("vector", memberType);
                break;
            default:
               throw new IllegalArgumentException("Unsupported Avro type: " + type);
        }

        return csmType;
    }

    public static JsonNode avroJsonToCsm(JsonNode typeNode) {
        ObjectNode csmType = objectMapper.createObjectNode();
         
        ObjectNode theField = (ObjectNode) typeNode.get("fields").get(0);
        String codeType = theField.get("name").asText();
        switch (codeType) {
            case "union": 
                var union = objectMapper.createObjectNode();
                var unionArray = (ArrayNode) theField.get("type");
                unionArray.forEach(variant -> {
                    ObjectNode field = (ObjectNode) variant.get("fields").get(0);
                    String variantName = fromHex(field.get("name").asText().substring(2));
                    union.put(variantName, field.get("type").asText());
                });
                csmType.set("union", union);
                break;
            case "product":
                var product = objectMapper.createObjectNode();
                var fields = (ArrayNode) theField.get("type").get("fields");
                fields.forEach(field -> {
                    String tagName = fromHex(field.get("name").asText().substring(2));
                    product.put(tagName, field.get("type").asText());
                });
                csmType.set("product", product);
                break;
            case "vector":
                String memberType = theField.get("type").get("fields").get(0).get("type").asText();
                csmType.put("vector", memberType);
                break;
            default:
                throw new IllegalArgumentException("Unsupported Avro type: " + codeType);
        }

        return csmType;
    }

    private static String toHex(String str) {
        StringBuilder encodedName = new StringBuilder();
        for (byte b : str.getBytes()) {
            encodedName.append(String.format("%02x", b));
        }
        return encodedName.toString();
    }

    private static String fromHex(String hexStr) {
        byte[] bytes = new byte[hexStr.length() / 2];
        for (int i = 0; i < bytes.length; i++) {
            bytes[i] = (byte) ((Character.digit(hexStr.charAt(2 * i), 16) << 4)
                    + Character.digit(hexStr.charAt(2 * i + 1), 16));
        }
        return new String(bytes);
    }
}

File: python_version/app/libs.py
from typing import Dict, Any, List
import json
from avro.schema import RecordSchema, Field, ArraySchema, UnionSchema

def to_hex(s: str) -> str:
    return ''.join([f'{b:02x}' for b in s.encode()])

def from_hex(hex_str: str) -> str:
    return bytes.fromhex(hex_str).decode()

def csm_to_avro(type_name: str, csm_type: Dict[str, Any]) -> Dict[str, Any]:
    avro_schema = {
        "name": type_name,
        "type": "record",
        "fields": [{
            "name": "union" if "union" in csm_type else "product" if "product" in csm_type else "vector",
            "type": None  # Will be set based on the type
        }]
    }
    
    if "union" in csm_type:
        union_types = []
        for tag_name, tag_type in csm_type["union"].items():
            union_type_record = {
                "name": f"{type_name}_{to_hex(tag_name)}",
                "type": "record",
                "fields": [{
                    "name": f"H_{to_hex(tag_name)}",
                    "type": tag_type
                }]
            }
            union_types.append(union_type_record)
        avro_schema["fields"][0]["type"] = union_types
        
    elif "product" in csm_type:
        product_type = {
            "name": f"{type_name}_record",
            "type": "record",
            "fields": []
        }
        
        for tag_name, tag_type in csm_type["product"].items():
            product_type["fields"].append({
                "name": f"H_{to_hex(tag_name)}",
                "type": tag_type
            })
        
        avro_schema["fields"][0]["type"] = product_type
        
    elif "vector" in csm_type:
        vector_type = {
            "name": f"{type_name}_vector",
            "type": "record",
            "fields": [{
                "name": "member",
                "type": csm_type["vector"]
            }]
        }
        avro_schema["fields"][0]["type"] = vector_type
        
    else:
        raise ValueError(f"Invalid CSM code: {csm_type}")
    
    return avro_schema

def avro_to_csm(type_schema: RecordSchema) -> Dict[str, Any]:
    the_field = type_schema['fields'][0]
    csm_type = {}
    if the_field['name'] == "union":
        union = {}
        for variant in the_field['type']:
            field = variant['fields'][0]
            variant_name = from_hex(field['name'][2:])
            union[variant_name] = field['type']
        csm_type["union"] = union
        
    elif the_field['name'] == "product":
        product = {}
        for field in the_field['type']['fields']:
            tag_name = from_hex(field['name'][2:])
            product[tag_name] = field['type']
        csm_type["product"] = product
        
    elif the_field['name'] == "vector":
        member_type = the_field['type']['fields'][0]['type']
        csm_type["vector"] = member_type
        
    else:
        raise ValueError(f"Unsupported Avro type: {type_schema}")
    
    return csm_type

def avro_json_to_csm(type_node: Dict[str, Any]) -> Dict[str, Any]:
    csm_type = {}
    the_field = type_node["fields"][0]
    code_type = the_field["name"]
    
    if code_type == "union":
        union = {}
        for variant in the_field["type"]:
            field = variant["fields"][0]
            variant_name = from_hex(field["name"][2:])
            union[variant_name] = field["type"]
        csm_type["union"] = union
        
    elif code_type == "product":
        product = {}
        for field in the_field["type"]["fields"]:
            tag_name = from_hex(field["name"][2:])
            product[tag_name] = field["type"]
        csm_type["product"] = product
        
    elif code_type == "vector":
        member_type = the_field["type"]["fields"][0]["type"]
        csm_type["vector"] = member_type
        
    else:
        raise ValueError(f"Unsupported Avro type: {code_type}")
    
    return csm_type 